dev/src/index.ts

```diff
@@ -1811,6 +1875,18 @@ export class Firestore implements firestore.Firestore {
                 'Received response: %j',
                 chunk
               );
+              numResponses++;
+              if (numResponses === 1) {
+                this._traceUtil
+                  .currentSpan()
+                  .addEvent(`Firestore.${methodName}: First response received`);
+              } else if (numResponses % NUM_RESPONSES_PER_TRACE_EVENT === 0) {
```

Comment:
How does tracing records snapshot listeners? Here it could capture the "first response" event, but could it get an "end" event? is it going to keep recording till the snapshot listeners called unsubscribe function?

Reply:
This is a great question. This function (`requestStream`) is called from 2 kinds of functions:

1- For server requests that require streaming RPCs. For example: when running `query.get()`. In these cases, the SDK creates a new span in the higher-level API function ([example](https://github.com/googleapis/nodejs-firestore/blob/4d9d4cbb3eab461dfc2250197a4ff9c198e6d3cc/dev/src/reference/query.ts#L1081)). So when `requestStream` is called, an active span exists already. Notice that inside `requestStream` we don't create a new span; we just say "get me the current span" (`.currentSpan()`) and add events to the current span. In these cases the SDK initiates the stream, sends the request, and when it receives the `end` event, it shuts down the stream. And then closes the _span_ in the higher level API function.

2- For snapshot listeners ([here](https://github.com/googleapis/nodejs-firestore/blob/f73e28b69be3ff9f870055af80683f12048549e8/dev/src/watch.ts#L484C12-L484C25)). In this case the higher level function that is calling `requestStream` is _not_ creating a span. So when `.currentSpan()` is called on line 1881, it calls [this](https://github.com/googleapis/nodejs-firestore/blob/4d9d4cbb3eab461dfc2250197a4ff9c198e6d3cc/dev/src/telemetry/enabled-trace-util.ts#L101), which calls `trace.getActiveSpan()` which will return `undefined`, so EnabledTraceUtil will create `new Span(undefined);`, which is a [valid](https://github.com/googleapis/nodejs-firestore/blob/4d9d4cbb3eab461dfc2250197a4ff9c198e6d3cc/dev/src/telemetry/span.ts#L21) usecase (note that the constructor parameter is optional). Then [`addEvent`](https://github.com/googleapis/nodejs-firestore/blob/4d9d4cbb3eab461dfc2250197a4ff9c198e6d3cc/dev/src/telemetry/span.ts#L28) will not do anything (it's a no-op) because there's no underlying otel span. So, for snapshot listeners, we do not create any spans and we do not add any events to any spans. This is intentional and consistent with the java implementation. Snapshot listeners are long-lived streams which are not the right fit for tracing; although we may define specific short-lived traces that raise visibility to some aspects of a snapshot listener in the future.